chia/wallet/nft_wallet/nft_off_chain.py

```diff
@@ -0,0 +1,96 @@
+import logging
+from pathlib import Path
+from typing import Optional
+
+import requests
+
+from chia.types.blockchain_format.program import Program
+from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.util.default_root import DEFAULT_ROOT_PATH
+from chia.util.hash import std_hash
+from chia.wallet.nft_wallet.nft_info import NFTCoinInfo
+from chia.wallet.nft_wallet.uncurry_nft import UncurriedNFT
+
+PREFIX_HASH_LENGTH = 3
+CACHE_PATH_KEY = "nft_metadata_cache_path"
+DEFAULT_PATH = DEFAULT_ROOT_PATH / "nft_cache"
+DEFAULT_PATH.mkdir(parents=True, exist_ok=True)
+log = logging.getLogger(__name__)
+
+
+def fetch_off_chain_metadata(nft_coin_info: NFTCoinInfo) -> Optional[str]:
+    uncurried_nft: Optional[UncurriedNFT] = UncurriedNFT.uncurry(*nft_coin_info.full_puzzle.uncurry())
+    if uncurried_nft is None:
+        return None
+    for uri in uncurried_nft.meta_uris.as_python():  # pylint: disable=E1133
+        response = requests.get(uri)
+        if response.status_code == 200:
+            return response.text
+    return None
+
+
+def read_off_chain_metadata(nft_coin_info: NFTCoinInfo, cache_path: Optional[str] = None) -> Optional[str]:
+    # Read metadata from disk cache
+    cache = DEFAULT_PATH
+    if cache_path is not None:
+        cache = Path(cache_path)
+    file_name = cache / nft_coin_info.nft_id.hex()[:PREFIX_HASH_LENGTH] / nft_coin_info.nft_id.hex()
+    if not file_name.exists():
+        return None
+    try:
+        text = file_name.read_text()
```

Comment:
Don't bother pre-checking.  You will still need to sensibly handle the exception since there's a race condition here where the file could get deleted between the existence check and the read attempt.